File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    currM = 0
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.func(root)
        return self.currM
        
    def func(self,r):
        if not r : return 0
        fl = self.func(r.left)
        fr = self.func(r.right)
        temp = fl + fr  
        if temp > self.currM : self.currM = temp
        return max(fl,fr) + 1
        

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = 0;
    int diameterOfBinaryTree(TreeNode* root) {
        helper(root);
        return ans;
    }
    int helper(TreeNode *node){
        if(node == nullptr) return 0;
        int left = helper(node->left);
        int right = helper(node->right);
        ans = max(ans, left+right);
        return max(left, right)+1;
    }
};